File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();

        for (String str: strs) {
            char[] alphabets = new char[26];
            for (char c: str.toCharArray()) {
                alphabets[c - 'a']++;
            }
            String key = new String(alphabets);
            map.computeIfAbsent(key, k -> new LinkedList<>());
            map.get(key).add(str);
        }

        return new LinkedList<>(map.values());
    }
}

File: 0049-group-anagrams/0049-group-anagrams.kt
class Solution {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        val alphabetSize = 26;
        val result: MutableMap<String, MutableList<String>> = mutableMapOf();

        for (str in strs) {
            val charCount = IntArray(alphabetSize);
            for (c in str) {
                charCount[c - 'a']++;
            }
            val key = charCount.joinToString(",");

            if (!result.containsKey(key)) {
                result[key] = mutableListOf()
            }
            result[key]?.add(str)
        }

        return result.values.toList();
    }
}